flocker/node/test/test_loop.py

```diff
@@ -0,0 +1,583 @@
+        loop = build_convergence_loop_fsm(deployer)
+        loop.receive(_ClientStatusUpdate(client=object(),
+                                         configuration=object(),
+                                         state=object()))
+        self.assertEqual(len(deployer.local_states), 0)  # Discovery started
+
+    def successful_amp_client(self, local_states):
+        """
+        Create AMP client that can respond successfully to a
+        ``NodeStateCommand``.
+
+        :param local_states: The node states we expect to be able to send.
+
+        :return FakeAMPClient: Fake AMP client appropriately setup.
+        """
+        client = FakeAMPClient()
+        for local_state in local_states:
+            client.register_response(
+                NodeStateCommand, dict(node_state=local_state),
+                {"result": None})
+        return client
+
+    def test_convergence_done_notify(self):
+        """
+        A FSM doing convergence that gets a discovery result notifies the last
+        received client.
+        """
+        local_state = object()
+        client = self.successful_amp_client([local_state])
+        action = ControllableAction(Deferred())
+        deployer = ControllableDeployer([succeed(local_state)], [action])
+        loop = build_convergence_loop_fsm(deployer)
+        loop.receive(_ClientStatusUpdate(client=client,
+                                         configuration=object(),
+                                         state=object()))
+        self.assertEqual(client.calls, [(NodeStateCommand,
+                                         dict(node_state=local_state))])
+
+    def test_convergence_done_changes(self):
+        """
+        A FSM doing convergence that gets a discovery result starts applying
+        calculated changes using last received desired configuration and
+        cluster state.
+        """
+        local_state = object()
+        configuration = object()
+        state = object()
+        action = ControllableAction(Deferred())
+        deployer = ControllableDeployer([succeed(local_state)], [action])
+        loop = build_convergence_loop_fsm(deployer)
+        loop.receive(_ClientStatusUpdate(
+            client=self.successful_amp_client([local_state]),
+            configuration=configuration, state=state))
+        # Calculating actions happened, and result was run:
+        self.assertEqual((deployer.calculate_inputs, action.called),
+                         ([(local_state, configuration, state)], True))
+
+    def test_convergence_done_start_new_iteration(self):
+        """
+        A FSM doing a convergence iteration does another iteration when
+        applying changes is done.
+        """
+        local_state = object()
+        local_state2 = object()
+        configuration = object()
+        state = object()
+        action = ControllableAction(succeed(None))
+        action2 = ControllableAction(Deferred())
```

Comment:
:exclamation:  It isn't at all obvious that argument here controls whether this action ends up running or not. It took a while of staring at the code to figure out why this tests runs two loops and the above test only runs one.